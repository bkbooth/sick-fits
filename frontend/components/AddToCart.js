import React from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';

export const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART($itemId: ID!) {
    addToCart(itemId: $itemId) {
      id
      quantity
    }
  }
`;

const AddToCart = ({ itemId }) => (
  <Mutation mutation={ADD_TO_CART_MUTATION} variables={{ itemId }}>
    {(addToCart, { loading }) => (
      <button onClick={addToCart} disabled={loading}>
        ðŸ›’ Add{loading ? 'ing' : ''} to cart
      </button>
    )}
  </Mutation>
);

AddToCart.propTypes = {
  itemId: PropTypes.string.isRequired,
};

export default AddToCart;
