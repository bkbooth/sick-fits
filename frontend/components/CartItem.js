import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import RemoveFromCart from './RemoveFromCart'
import formatMoney from '../lib/formatMoney'

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid ${props => props.theme.lightgrey};
  display: grid;
  align-items: center;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 10px;
  }
  h3,
  p {
    margin: 0;
  }
`

const CartItem = ({ cartItem }) => {
  if (!cartItem.item) return (
    <CartItemStyles>
      <p>This item has been removed.</p>
      <RemoveFromCart id={cartItem.id} />
    </CartItemStyles>
  )
  return <CartItemStyles>
    <img src={cartItem.item.image} width="100" alt={cartItem.item.title} />
    <div className="cart-item-details">
      <h3>{cartItem.item.title}</h3>
      <p>
        {formatMoney(cartItem.item.price * cartItem.quantity)}
        {' - '}
        <em>{cartItem.quantity} Ã— {formatMoney(cartItem.item.price)} each</em>
      </p>
    </div>
    <RemoveFromCart id={cartItem.id} />
  </CartItemStyles>
}

CartItem.propTypes = {
  cartItem: PropTypes.shape({
    id: PropTypes.string,
    quantity: PropTypes.number,
    item: PropTypes.shape({
      id: PropTypes.string,
      title: PropTypes.string,
      description: PropTypes.string,
      image: PropTypes.string,
      price: PropTypes.number,
    })
  }).isRequired,
}

export default CartItem
