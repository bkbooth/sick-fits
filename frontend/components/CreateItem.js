import React, { Component } from 'react'
import Router from 'next/router'
import { Mutation } from 'react-apollo'
import gql from 'graphql-tag'
import Form from './styles/Form'
import formatMoney from '../lib/formatMoney'
import Error from './ErrorMessage'

const CREATE_ITEM_MUTATION = gql`
  mutation CREATE_ITEM_MUTATION(
    $title: String!
    $description: String!
    $price: Int!
    $image: String
    $largeImage: String
  ) {
    createItem(
      title: $title
      description: $description
      price: $price
      image: $image
      largeImage: $largeImage
    ) {
      id
    }
  }
`

class CreateItem extends Component {
  state = {
    title: '',
    description: '',
    image: '',
    largeImage: '',
    price: 0,
  }

  handleChange = e => {
    const { name, type, value } = e.target
    const val = type === 'number' ? parseFloat(value) : value
    this.setState({ [name]: val })
  }

  uploadFile = async e => {
    const files = e.target.files
    const data = new FormData()
    data.append('file', files[0])
    data.append('upload_preset', 'sick-fits')

    const res = await fetch('https://api.cloudinary.com/v1_1/bkbooth/image/upload', {
      method: 'POST',
      body: data,
    })
    const file = await res.json()

    this.setState({
      image: file.secure_url,
      largeImage: file.eager[0].secure_url,
    })
  }

  render() {
    return (
      <Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
        {(createItem, { loading, error }) => (
          <Form
              onSubmit={async e => {
              // Stop the form from submitting
              e.preventDefault()
              // Call the mutation
              const res = await createItem()
              // Redirect user to the single item page
              Router.push({
                pathname: '/item',
                query: { id: res.data.createItem.id },
              })
            }}
            data-test="form"
          >
            <Error error={error} />

            <fieldset disabled={loading} aria-busy={loading}>
              <label htmlFor="file">
                Image
                <input
                  type="file"
                  id="file"
                  name="file"
                  placeholder="Upload an image"
                  required
                  onChange={this.uploadFile}
                />
                {this.state.image && <img src={this.state.image} width="200" alt="Upload Preview" />}
              </label>

              <label htmlFor="title">
                Title
                <input
                  type="text"
                  id="title"
                  name="title"
                  placeholder="Title"
                  required
                  value={this.state.title}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="price">
                Price
                <input
                  type="number"
                  id="price"
                  name="price"
                  placeholder="Price"
                  required
                  value={this.state.price}
                  onChange={this.handleChange}
                />
              </label>

              <label htmlFor="description">
                Description
                <textarea
                  id="description"
                  name="description"
                  placeholder="Enter a description"
                  required
                  value={this.state.description}
                  onChange={this.handleChange}
                />
              </label>

              <button type="submit">Submit</button>
            </fieldset>
          </Form>
        )}
      </Mutation>
    )
  }
}

export default CreateItem
export { CREATE_ITEM_MUTATION }
