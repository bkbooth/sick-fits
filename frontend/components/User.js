import { useQuery } from '@apollo/react-hooks';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';

export const CURRENT_USER_QUERY = gql`
  query CURRENT_USER {
    me {
      id
      email
      name
      permissions
      cart {
        id
        quantity
        item {
          id
          title
          price
          image
        }
      }
    }
  }
`;

const User = ({ children }) => {
  const payload = useQuery(CURRENT_USER_QUERY);
  return children(payload);
};

User.propTypes = {
  children: PropTypes.func.isRequired,
};

export default User;
