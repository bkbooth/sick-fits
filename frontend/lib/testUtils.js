import casual from 'casual';

// seed it so we get consistent results
casual.seed(777);

export const fakeUser = () => ({
  __typename: 'User',
  id: casual.uuid,
  name: casual.name,
  email: casual.email,
  permissions: ['ADMIN'],
  orders: [],
  cart: [],
});

export const fakeItem = () => ({
  __typename: 'Item',
  id: casual.uuid,
  title: casual.title,
  description: casual.description,
  price: casual.integer(1000, 30000),
  image: `${casual.word}.jpg`,
  largeImage: `${casual.word}-large.jpg`,
  user: null,
});

export const fakeCartItem = overrides => ({
  __typename: 'CartItem',
  id: casual.uuid,
  quantity: casual.integer(1, 5),
  item: fakeItem(),
  user: fakeUser(),
  ...overrides,
});

export const fakeOrderItem = () => ({
  __typename: 'OrderItem',
  id: casual.uuid,
  title: casual.title,
  description: casual.description,
  price: casual.integer(1000, 30000),
  image: `${casual.word}.jpg`,
  quantity: casual.integer(1, 5),
});

export const fakeOrder = () => ({
  __typename: 'Order',
  id: casual.uuid,
  charge: `ch_${casual.uuid.split('-')[4]}`,
  total: casual.integer(10000, 100000),
  items: [fakeOrderItem(), fakeOrderItem()],
  createdAt: casual.date('YYYY-MM-DD[T]HH:mm:ss[Z]'),
  user: fakeUser(),
});

export function typeInto(wrapper, { inputType = 'input', name, type = 'text', value }) {
  wrapper
    .find(`${inputType}[name="${name}"]`)
    .simulate('change', { target: { name, type, value } });
}
