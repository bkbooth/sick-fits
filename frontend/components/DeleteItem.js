import React from 'react';
import { Mutation } from 'react-apollo';
import gql from 'graphql-tag';
import PropTypes from 'prop-types';
import { ALL_ITEMS_QUERY } from 'components/Items';

export const DELETE_ITEM_MUTATION = gql`
  mutation DELETE_ITEM($itemId: ID!) {
    deleteItem(id: $itemId) {
      id
    }
  }
`;

const DeleteItem = ({ itemId }) => {
  function createDeleteHandler(mutation) {
    return () =>
      confirm('Are you sure you want to delete this item?') &&
      mutation().catch(error => alert(error.message));
  }

  function cacheUpdate(cache, payload) {
    const data = cache.readQuery({ query: ALL_ITEMS_QUERY });
    data.items = data.items.filter(item => item.id !== payload.data.deleteItem.id);
    cache.writeQuery({ query: ALL_ITEMS_QUERY, data });
  }

  return (
    <Mutation mutation={DELETE_ITEM_MUTATION} variables={{ itemId }} update={cacheUpdate}>
      {(deleteItem, { loading }) => (
        <button onClick={createDeleteHandler(deleteItem)} disabled={loading}>
          🗑️ Delet{loading ? 'ing' : 'e'} item
        </button>
      )}
    </Mutation>
  );
};

DeleteItem.propTypes = {
  itemId: PropTypes.string.isRequired,
};

export default DeleteItem;
