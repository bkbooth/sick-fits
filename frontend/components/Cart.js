import React from 'react'
import { Query, Mutation } from 'react-apollo'
import { adopt } from 'react-adopt'
import gql from 'graphql-tag'
import CartStyles from './styles/CartStyles'
import Supreme from './styles/Supreme'
import CloseButton from './styles/CloseButton'
import SickButton from './styles/SickButton'
import User from './User'
import CartItem from './CartItem'
import TakeMyMoney from './TakeMyMoney'
import calcTotalPrice from '../lib/calcTotalPrice'
import formatMoney from '../lib/formatMoney'
import apostrophiseName from '../lib/apostrophiseName'

const LOCAL_STATE_QUERY = gql`
  query LOCAL_STATE_QUERY {
    cartOpen @client
  }
`

const TOGGLE_CART_MUTATION = gql`
  mutation TOGGLE_CART_MUTATION {
    toggleCart @client
  }
`

const Composed = adopt({
  user: ({ render }) => <User>{render}</User>,
  toggleCart: ({ render }) => <Mutation mutation={TOGGLE_CART_MUTATION}>{render}</Mutation>,
  localState: ({ render }) => <Query query={LOCAL_STATE_QUERY}>{render}</Query>,
})

const Cart = () => (
  <Composed>
    {({ user, toggleCart, localState }) => {
      const me = user.data.me
      if (!me) return null
      return (
        <CartStyles open={localState.data.cartOpen}>
          <header>
            <CloseButton
              title="close"
              onClick={toggleCart}
            >Ã—</CloseButton>
            <Supreme>{apostrophiseName(me.name)} Cart</Supreme>
            <p>You have {me.cart.length} item{me.cart.length === 1 ? '': 's'} in your cart.</p>
          </header>

          <ul>
            {me.cart.map(cartItem => <CartItem cartItem={cartItem} key={cartItem.id} />)}
          </ul>

          <footer>
            <p>{formatMoney(calcTotalPrice(me.cart))}</p>
            <TakeMyMoney>
              <SickButton disabled={!me.cart.length}>Checkout</SickButton>
            </TakeMyMoney>
          </footer>
        </CartStyles>
      )
    }}
  </Composed>
)

export default Cart
export { LOCAL_STATE_QUERY, TOGGLE_CART_MUTATION }
