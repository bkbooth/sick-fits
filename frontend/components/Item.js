import React from 'react';
import Link from 'next/link';
import PropTypes from 'prop-types';
import formatMoney from 'lib/formatMoney';
import DeleteItem from 'components/DeleteItem';
import ItemStyles from 'components/styles/ItemStyles';
import PriceTag from 'components/styles/PriceTag';
import Title from 'components/styles/Title';

const Item = ({ item }) => (
  <ItemStyles>
    {item.image && <img src={item.image} alt={item.title} />}

    <Title>
      <Link href="/items/[itemId]" as={`/items/${item.id}`}>
        <a>{item.title}</a>
      </Link>
    </Title>
    <PriceTag>{formatMoney(item.price)}</PriceTag>
    <p>{item.description}</p>

    <div className="buttonList">
      <Link href="/items/[itemId]/update" as={`/items/${item.id}/update`}>
        <a>✏️ Edit</a>
      </Link>
      <button>Add to cart</button>
      <DeleteItem itemId={item.id}>Delete</DeleteItem>
    </div>
  </ItemStyles>
);

Item.propTypes = {
  item: PropTypes.shape({
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    price: PropTypes.number.isRequired,
    image: PropTypes.string,
  }),
};

export default Item;
