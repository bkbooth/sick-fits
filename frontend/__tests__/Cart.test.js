import { mount } from 'enzyme'
import toJSON from 'enzyme-to-json'
import wait from 'waait'
import { MockedProvider } from 'react-apollo/test-utils'
import { fakeUser, fakeCartItem } from '../lib/testUtils'
import { CURRENT_USER_QUERY } from '../components/User'
import Cart, { LOCAL_STATE_QUERY } from '../components/Cart'

const mocks = [{
  request: { query: CURRENT_USER_QUERY },
  result: {
    data: {
      me: {
        ...fakeUser(),
        cart: [fakeCartItem()],
      },
    },
  },
}, {
  request: { query: LOCAL_STATE_QUERY },
  result: { data: { cartOpen: true } },
}]

describe('<Cart />', () => {
  it('renders and matches snapshot', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <Cart />
      </MockedProvider>
    )
    await wait()
    wrapper.update()
    expect(toJSON(wrapper.find('header'))).toMatchSnapshot()
    expect(wrapper.find('CartItem')).toHaveLength(1)
  })
})
