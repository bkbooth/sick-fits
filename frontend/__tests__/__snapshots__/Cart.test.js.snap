// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> renders and matches snapshot 1`] = `
<header>
  <CloseButton
    onClick={[Function]}
    title="close"
  >
    <StyledComponent
      forwardedClass={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "CloseButton-sc-1seb878-0",
            "isStatic": true,
            "lastClassName": "lbNKfp",
            "rules": Array [
              "background:black;color:white;font-size:3rem;border:0;position:absolute;z-index:2;right:0;",
            ],
          },
          "displayName": "CloseButton",
          "render": [Function],
          "styledComponentId": "CloseButton-sc-1seb878-0",
          "target": "button",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
      onClick={[Function]}
      title="close"
    >
      <button
        className="CloseButton-sc-1seb878-0 lbNKfp"
        onClick={[Function]}
        title="close"
      >
        Ã—
      </button>
    </StyledComponent>
  </CloseButton>
  <Supreme>
    <StyledComponent
      forwardedClass={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "attrs": Array [],
          "componentStyle": ComponentStyle {
            "componentId": "Supreme-xv30qb-0",
            "isStatic": false,
            "lastClassName": "hpaXsq",
            "rules": Array [
              "background:",
              [Function],
              ";color:white;display:inline-block;padding:4px 5px;transform:skew(-3deg);margin:0;font-size:4rem;",
            ],
          },
          "displayName": "Supreme",
          "render": [Function],
          "styledComponentId": "Supreme-xv30qb-0",
          "target": "h3",
          "toString": [Function],
          "warnTooManyClasses": [Function],
          "withComponent": [Function],
        }
      }
      forwardedRef={null}
    >
      <h3
        className="Supreme-xv30qb-0 hpaXsq"
      >
        Miss Coleman Berge's
         Cart
      </h3>
    </StyledComponent>
  </Supreme>
  <p>
    You have 
    1
     item
     in your cart.
  </p>
</header>
`;
